modules/extensions-support/src/main/java/org/mule/module/extension/internal/runtime/config/DynamicConfigurationInstanceProvider.java

```diff
@@ -0,0 +1,180 @@
+     * @param resolverSet                the {@link ResolverSet} that's going to be evaluated
+     */
+    public DynamicConfigurationInstanceProvider(String name,
+                                                Extension extension,
+                                                ConfigurationInstanceRegistrationCallback registrationCallback,
+                                                ConfigurationObjectBuilder configurationObjectBuilder,
+                                                ResolverSet resolverSet,
+                                                ExpirationPolicy expirationPolicy)
+    {
+        this.name = name;
+        this.extension = extension;
+        this.registrationCallback = registrationCallback;
+        this.configurationObjectBuilder = configurationObjectBuilder;
+        this.resolverSet = resolverSet;
+        this.expirationPolicy = expirationPolicy;
+    }
+
+    /**
+     * Evaluates {@link #resolverSet} using the given {@code event} and returns
+     * an instance produced with the result. For equivalent {@link ResolverSetResult}s
+     * it will return the same instance.
+     *
+     * @param operationContext a {@link OperationContext}
+     * @return the resolved value
+     */
+    @Override
+    public T get(OperationContext operationContext)
+    {
+        try
+        {
+            ResolverSetResult result = resolverSet.resolve(asOperationContextAdapter(operationContext).getEvent());
+            DynamicConfigurationInstanceHolder configurationInstanceHolder = getHolder(result);
+
+            operationContext.onOperationSuccessful(signal -> discountUsage(configurationInstanceHolder));
+            operationContext.onOperationFailed(signal -> discountUsage(configurationInstanceHolder));
+
+            configurationInstanceHolder.accountUsage();
+            return (T) configurationInstanceHolder.getConfigurationInstance();
+        }
+        catch (Exception e)
+        {
+            throw new MuleRuntimeException(e);
+        }
+    }
+
+    private DynamicConfigurationInstanceHolder getHolder(ResolverSetResult result) throws Exception
+    {
+        DynamicConfigurationInstanceHolder configurationInstance;
+        cacheReadLock.lock();
+        try
+        {
+            configurationInstance = cache.get(result);
+            if (configurationInstance == null)
+            {
+                cacheReadLock.unlock();
+                cacheWriteLock.lock();
+
+                try
+                {
+                    // re-check in case some other thread beat us to it...
+                    configurationInstance = cache.get(result);
+                    if (configurationInstance == null)
+                    {
+                        configurationInstance = createAndRegisterConfigurationInstance(result);
+                        cache.put(result, configurationInstance);
+                    }
+
+                    // downgrade
+                    cacheReadLock.lock();
```

Comment:
This downgrade seems required only because the outer block of code is surrounded by a finally that is doing cacheReadLock.unlock(). I think the code would be more readable if it has tow parts, one using only the read lock to check for existence and another one using only the write lock